lib/routes/v1.js

```diff
@@ -59,6 +60,97 @@ let api = new base.API({
+
+  let reasons = [];
+
+  // First let's ensure that the launch specs are valid and able to be
+  // generated
+  let launchSpecs = [];
+  try {
+    launchSpecs = ctx.WorkerType.testLaunchSpecs(
+        workerType,
+        ctx.keyPrefix,
+        ctx.provisionerId,
+        ctx.provisionerBaseUrl,
+        ctx.pubKey,
+        workerType);
+    debug(`generated launch specifications for ${workerTypeName}`);
+  } catch (err) {
+    // We don't want to handle other things breaking, just badly formed worker
+    // types
+    debug(`could not generate launch specifications for ${workerTypeName}`);
+    if (err && err.code !== 'InvalidLaunchSpecifications') {
+      throw err;
+    }
+    reasons.concat(err.reasons);
+  }
+
+  await Promise.all(_.map(launchSpecs, async (region, regionName) => {
+    await Promise.all(_.map(region, async (type, typeName) => {
+      let launchSpec = type.launchSpec;
+      // We delete the Placement.AvailabilityZone key because we aren't don't
+      // know which AZ to test for and this shouldn't matter for these purposes
+      if (launchSpec.Placement && launchSpec.Placement.AvailabilityZone) {
+        delete launchSpec.Placement.AvailabilityZone;
+      }
+      // Next, let's check if the launch spec has valid parameters
+      try {
+        await ctx.ec2[regionName].requestSpotInstances({
+          DryRun: true,
+          InstanceCount: 1,
+          Type: 'one-time',
+          LaunchSpecification: launchSpec,
+          SpotPrice: '1', // Only since this is a DryRun:true call
+        }).promise();
+        debug(`did DryRun of requesting spot instance for ${workerTypeName}`);
+      } catch (err) {
+        if (err.code !== 'DryRunOperation') {
+          debug(`could not do a DryRun of requesting spot instance for ${workerTypeName}`);
+          reasons.push(err)
+        }
+      }
+      let images = [];
+      try {
+        images = await ctx.ec2[regionName].describeImages({
+          ImageIds: [launchSpec.ImageId],
+        }).promise();
+        if (images.data.Images.length !== 1) {
+          reasons.push(new Error(`Too many results found for ${launchSpec.ImageId}`)); 
+        }
+        let image = images.data.Images[0];
+        if (image.ImageId !== launchSpec.ImageId) {
+          reasons.push(new Error(`Image ID returned from EC2 api ${image.ImageId} does not match launch spec ${launchSpec.ImageId}`));
+        }
+        if (image.State !== 'available') {
+          reasons.push(new Error(`Image state for ${launchSpec.ImageId} must be 'available' not ${image.State}`));
+        }
+      } catch (err) {
+        reasons.push(err);
+      }
+    }));
+  }));
+
+  // Finally, let's verify that the image for this workerType exists in EC2
+
+  if (reasons.length > 0) {
+    debug('Found errors with ' + workerTypeName); 
+    let e = new Error('Refusing to create an invalid worker type');
+    e.reasons = reasons;
+    throw e
```

Comment:
@jhford, this okay... But really you shouldn't communicate it this way...

`validateWorkerType` should just return a string message if there is an error, or a `null` if there is no error..
Or something like this, you generally shouldn't use exceptions to communicate return value.